integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/common/BenchUtils.scala

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.nvidia.spark.rapids.tests.common
+
+import java.io.{File, FileOutputStream}
+import java.time.Instant
+import java.util.concurrent.TimeUnit.NANOSECONDS
+
+import scala.collection.mutable.ListBuffer
+
+import org.json4s.DefaultFormats
+import org.json4s.jackson.JsonMethods.parse
+import org.json4s.jackson.Serialization.writePretty
+
+import org.apache.spark.{SPARK_BUILD_USER, SPARK_VERSION}
+import org.apache.spark.sql.{DataFrame, Row, SparkSession}
+
+object BenchUtils {
+
+  /**
+   * Run the specified number of cold and hot runs and record the timings and summary of the
+   * query and results to file, including all Spark configuration options and environment
+   * variables.
+   *
+   * @param spark The Spark session
+   * @param queryRunner Function to create a DataFrame from the Spark session.
+   * @param queryDescription A query name to be written to the JSON file.
+   * @param filenameStub The prefix the for the JSON file. The current timestamp will be appended
+   *                     to ensure that filenames are unique and that results are not inadvertently
+   *                     overwritten.
+   * @param numColdRuns The number of cold runs.
+   * @param numHotRuns The number of hot runs.
+   * @param maxResultsToRecord The maximum number of result rows to collect and store in the
+   *                           output JSON file. This can be used as a sanity check to compare
+   *                           to previous runs but is not intended to be used for full results
+   *                           verification. Setting this to `None` will capture all result rows,
+   *                           and setting this to `Some(0)` will skip recording results.
+   */
+  def runBench(
+      spark: SparkSession,
+      queryRunner: SparkSession => DataFrame,
+      queryDescription: String,
+      filenameStub: String,
+      numColdRuns: Int,
+      numHotRuns: Int,
+      maxResultsToRecord: Option[Int]): Unit = {
+
+    val queryStartTime = Instant.now()
+
+    var df: DataFrame = null
+    var results: Seq[Row] = null
+    val coldRunElapsed = new ListBuffer[Long]()
+    for (i <- 0 until numColdRuns) {
```

Comment:
How do we ensure its a cold run in a loop? As I understand a cold run is the first time we run the query, so how is it a cold run in a loop? Same question for the hot run, how is it a hot run in a loop? May be my understanding of hot v cold is flawed

Reply:
This code is using the term "cold run" to mean any run that happens before we start to collect timing information that will be used for reporting final results. I agree that the first one is the only truly cold run. 

The goal of the hot run loop is to gather multiple timings so we can choose to pick the best one, or an average, when reporting results, since query times do vary a bit between runs.

This was the existing logic that was duplicated in TPC-DS and TPCxBB, and I just moved it into the new BenchUtil class in this PR, but I'm open to changing it. Perhaps we should just specify the total number of runs instead, and when reporting results we can choose whether to disregard the first N results if we are computing an average, or perhaps we just want to pick the best time out of all runs. What do you think @abellina ?

Reply:
Sorry, I didn't realize you just moved the code around. We can have this change as a follow-on if you prefer. 

We could pass that in as a parameter whether to pick the best time or average to start with. We can then expand on other stats later. 

For the cold run, is it possible to "reset" the session so we can have actual cold runs?

Reply:
Yeah I agree the loop there makes unclear what "cold" is. I think all we really need is to distinguish between the first run, and the others. It is useful to know how much overhead there is that first go around.

Reply:
I removed the two loops. We now just have `iterations`. I did leave separate reporting for cold/hot timings for the case where `iterations > 1`. Let me know if this is clearer.